src/backend/distributed/master/master_multi_shard_modify.c

```diff
@@ -0,0 +1,402 @@
+
+static int SendQueryToConnections(char *shardQueryString,
+								  ShardConnections *shardConnections);
+static void LockShardsForModify(List *shardIntervalList);
+PG_FUNCTION_INFO_V1(master_multi_shard_modify);
+
+
+/*
+ * master_multi_shard_modify takes in a delete query string and pushes the
+ * query to shards. It finds shards that match the criteria defined in the
+ * delete command, generates the same delete query string for each of the
+ * found shards with distributed table name replaced with the shard name and
+ * sends the queries to the workers. It uses one-phase or two-phase commit
+ * transactions depending on citus.copy_transaction_manager value.
+ */
+Datum
+master_multi_shard_modify(PG_FUNCTION_ARGS)
+{
+	text *queryText = PG_GETARG_TEXT_P(0);
+	char *queryString = text_to_cstring(queryText);
+	List *queryTreeList = NIL;
+	Oid relationId = InvalidOid;
+	Index tableId = 1;
+	Query *modifyQuery = NULL;
+	Node *queryTreeNode;
+	List *restrictClauseList = NIL;
+	bool isTopLevel = true;
+	bool failOK = false;
+	List *shardIntervalList = NIL;
+	List *prunedShardIntervalList = NIL;
+	int32 affectedTupleCount = 0;
+
+	PreventTransactionChain(isTopLevel, "master_multi_shard_modify");
+
+	queryTreeNode = ParseTreeNode(queryString);
+	if (IsA(queryTreeNode, DeleteStmt))
+	{
+		DeleteStmt *deleteStatement = (DeleteStmt *) queryTreeNode;
+		relationId = RangeVarGetRelid(deleteStatement->relation, NoLock, failOK);
+	}
+	else if(IsA(queryTreeNode, UpdateStmt))
+	{
+		UpdateStmt *updateStatement = (UpdateStmt *) queryTreeNode;
+		relationId = RangeVarGetRelid(updateStatement->relation, NoLock, failOK);
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("query \"%s\" is not a delete nor update statement",
+							   queryString)));
+	}
+
+	CheckDistributedTable(relationId);
+
+	queryTreeList = pg_analyze_and_rewrite(queryTreeNode, queryString, NULL, 0);
+	modifyQuery = (Query *) linitial(queryTreeList);
+
+	ErrorIfModifyQueryNotSupported(modifyQuery);
+
+	shardIntervalList = LoadShardIntervalList(relationId);
+	restrictClauseList = WhereClauseList(modifyQuery->jointree);
+
+	prunedShardIntervalList =
+		PruneShardList(relationId, tableId, restrictClauseList, shardIntervalList);
+	
+	LockShardsForModify(prunedShardIntervalList);
+
+	affectedTupleCount = SendQueryToShards(modifyQuery, prunedShardIntervalList);
+
+	PG_RETURN_INT32(affectedTupleCount);
+}
+
+
+/*
+ * LockShardsForModify command locks the replicas of given shard. The
+ * lock logic is slightly different from LockShards function. Basically,
+ * 
+ * 1. If citus.all_modifications_commutative is set to true, then all locks
+ * are acquired as ShareLock.
+ * 2. If citus.all_modifications_commutative is false, then only the shards
+ * with 2 or more replicas are locked with ExclusiveLock. Otherwise, the 
+ * lock is acquired with ShareLock.
+ */
+static void
+LockShardsForModify(List *shardIntervalList)
+{
+	if (AllModificationsCommutative)
+	{
+		LockShards(shardIntervalList, ShareLock);
```

Comment:
These should also be in order to avoid deadlock (e.g. transaction A takes exclusive lock on shard 1, transaction B takes share lock on shard 2, transaction A tries to get exclusive lock on shard 2 but blocks, transaction B tries to get share lock on shard 1 but blocks - deadlock). This situation may not currently exists, since nothing else takes multiple exclusive locks if AllModificationsCommutative is true, but that's more of a coincidence that we might easily miss in the future. I think you can just move the AllModificationsCommutative check down into the loop.